### ISSUE TITLE
Occasional freeze during "Finishing operation..."

### ISSUE THREAD

**Issue by Malivil** [axodox/AxoTools#30]
Occasionally (and it may only be after a test fails in the previous run), the coverage will freeze at the "Finishing operation..." step.

The output shows in 1 of 2 ways when it freezes.

Freeze 1:
> Test execution started.
> INFO Loading assembly from C:\Users\Malivil\AppData\Local\Microsoft\VisualStudio\14.0\Extensions\oudpwur2.hv0\NUnitAdapter\NUnit3.TestAdapter.dll...
> INFO Assembly loaded.
> INFO Loading assembly from C:\Users\Malivil\AppData\Local\Microsoft\VisualStudio\14.0\Extensions\oudpwur2.hv0\xUnitAdapter\xunit.runner.visualstudio.testadapter.dll...
> INFO Assembly loaded.
> INFO Loading assembly from C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\IDE\CommonExtensions\Microsoft\TestWindow\Extensions\Microsoft.VisualStudio.TestPlatform.Extensions.VSTestIntegration.dll...
> INFO Assembly loaded.
> INFO Executing tests...
> INFO Running executor: Microsoft.VisualStudio.TestPlatform.Extensions.VSTestIntegration.MSTestExecutor...
> INFO Passed   [Insert all my tests here]

Freeze 2:
> Test execution started.
> INFO Loading assembly from C:\Users\Malivil\AppData\Local\Microsoft\VisualStudio\14.0\Extensions\oudpwur2.hv0\NUnitAdapter\NUnit3.TestAdapter.dll...
> INFO Assembly loaded.
> INFO Loading assembly from C:\Users\Malivil\AppData\Local\Microsoft\VisualStudio\14.0\Extensions\oudpwur2.hv0\xUnitAdapter\xunit.runner.visualstudio.testadapter.dll...
> INFO Assembly loaded.
> INFO Loading assembly from C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\IDE\CommonExtensions\Microsoft\TestWindow\Extensions\Microsoft.VisualStudio.TestPlatform.Extensions.VSTestIntegration.dll...
> INFO Assembly loaded.
> INFO Executing tests...
> INFO Running executor: Microsoft.VisualStudio.TestPlatform.Extensions.VSTestIntegration.MSTestExecutor...
> INFO Passed   [Insert all my tests here]
> INFO Executor finished.
> INFO Test execution finished.

AxoCover.Host-x86.exe and OpenCover.Console.exe are still running, but are sitting at 0% CPU and unchanging memory usage. If I force close them (or click "Abort") and the vstest.*.exe programs and re-run the coverage, it usually works at least once again.

I haven't tried downgrading to the version I had been using, but I may do that in a bit to see if it helps.

Versions:
Visual Studio 2015 Professional
AxoCover 1.1.0.46 (build 1.1.49)

**Comment by axodox** [comment]
For me this usually happens after debugging tests, can you confirm if for you it happens even without that? 

(I am still to fix that issue, but nowadays I am fighting with the MSTestV2 adapter of Microsoft, as loading it cause me a headache...)

If you attach a debugger, can you take a loot at the stack?

**Comment by Malivil** [comment]
I was doing a lot of test debugging at the time and now that I haven't debugged a test in a while I also haven't experienced this freeze.

Do I need the Visual Studio Addins SDK installed to debug and what executable should I attach to? AxoCover.Host-x86.exe?

**Comment by axodox** [comment]
Yes the extension SDK is needed, but as I know there is a NuGet package called Microsoft.VSSDK.BuildTools to build without it - maybe you can try and add it to project AxoCover.

The UI part of AxoCover and the AxoCover.Host-*.exe processes are communicating through a TCP session. When you debug the test, the connection will timeout when breaking into the code, however the test host will cache all results and there is a mechanism to rebuild the connection in the end to transfer them to the UI, but sometimes recovery seem to fail for some reason.

If you debug I suggest you to start from the class AxoTestRunner, and as you go into the invoked code you will see how the connection is managed. From there we need to figure out why the connection fails to recover after debugging. While the issue does not happen after all test runs with a debugging break, it can be reproduced with high frequency, so you should not have much issue with reproducing the issue during debugging.

**Comment by Malivil** [comment]
Alright, I will try that out and let you know how it goes

**Comment by axodox** [comment]
Thank you. I work on AxoCover alone in my free time, so all help is greatly appreciated.

**Comment by Malivil** [comment]
I ran through a few thousand tests (yes, really), combining debugging through AxoCover, debugging through MSTest, tests failing, etc. and was not able to reproduce the problem. I am currently using 1.1.0.47 (build 1.1.54) which is newer than the version I was experiencing this with, so it's possible it was fixed by something else?

Either way, I now know how to debug it so if the problem does come back I can dig into it at that time.

**Comment by axodox** [comment]
Thanks for your effort. While there were some fixes, as far as I know the issue still happens in some cases. I will investigate it once I finish with the MSTestV2 issues.

**Comment by axodox** [comment]
I think I found a possible source of the said issue: it seems that if tests are creating foreground threads and they are left running they can prevent the test runner from closing, which then keeps OpenCover waiting. In my case this happens occasionally, and results in being stuck in "Finishing operation..." state, which might be the same issue you have encountered, I will push a fix and we see if it helps.

I have also recently made some architectural changes which makes the threading / networking related to test runner much simpler and more robust, I hope this will resolve instability problems.
